### PR TITLE
Logging Refactoring

### PR DESCRIPTION
**Pull request by ykonyshev**
Implemented a global logging configuration as well as child loggers for modules. Replaced some `print`calls and `traceback.print_exc` with corresponding `logger.*` calls.
Separated out the argument parsing from the `tagstudio/tagstudio.py` to `tagstudio/parse_args.py`.
Also ran `ruff check --fix` and added the [ruff-pre-commit](https://github.com/astral-sh/ruff-pre-commit) pre-commit hook. Refactored the code accordignly to satisfy the linter.

### CODE REVIEW

**Review by Loran425 — COMMENTED** [review]
Definitely want @CyanVoxel's input on the match/case in tag_studio.py and the possibility of removing the `ui` folder, and I think there's some room to discuss the log formatter and what information should be displayed as well as some questions on the ContextVars and splitting out the arg parser, though that split has been growing on me. 

Nice Job!

**Comment by MrSuperTop** [comment]
@Loran425 Thank you for the review. I will have a look at the comments later, most likely tomorrow.

File: tagstudio/tag_studio.py [hunk 1]
```diff
@@ -4,83 +4,46 @@
...
+    match args.ui:
+        case "qt":
+            driver = QtDriver(core, args)
+            ui_name = "Qt"
+        case "cli":
+            driver = CliDriver(core, args)
+            ui_name = "CLI"
+        case _:
+            driver = QtDriver(core, args)
+            ui_name = "Qt"
```
[the remark this answers is not in the document]
Comment (CyanVoxel) [reply tagstudio/tag_studio.py:38]: I'm all for the match/case statements, completely forgot that was added. 3.9 support is already definitely dropped so this should be find moving forward 👍

File: tagstudio/src/qt/ts_qt.py [hunk 2]
```diff
@@ -244,7 +223,7 @@ def setup_signals(self):
     def start(self) -> None:
         """Launches the main Qt window."""
 
-        loader = QUiLoader()
+        # loader = QUiLoader()
```
[the remark this answers is not in the document]
Comment (CyanVoxel) [reply tagstudio/src/qt/ts_qt.py:226]: It's indeed a holdover, good catch

File: tagstudio/src/core/logging.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+_main_logger = _root_logger.getChild("tag_studio")
+
+_main_logger_context_var: ContextVar[logging.Logger] = ContextVar("main_logger")
+_main_logger_context_var.set(_main_logger)
+
+
+def setup_logging() -> None:
+    logging.basicConfig(
+        format="%(asctime)s | %(levelname)s: %(message)s", level=logging.DEBUG
+    )
```
[the remark this answers is not in the document]
Comment (CyanVoxel) [reply tagstudio/src/core/logging.py:14]: Love this idea, I was thinking the same thing

**Comment by CyanVoxel** [comment]
> Definitely want @CyanVoxel's input on the match/case in tag_studio.py and the possibility of removing the `ui` folder, and I think there's some room to discuss the log formatter and what information should be displayed as well as some questions on the ContextVars and splitting out the arg parser, though that split has been growing on me.

It would probably be good to scrap the UI folder at this point. I think the UI development is pretty solidified in just Python at this point, for better or for worse. And I commented above but just in case it gets missed, I'm all for adding in match/case statements!

Thanks so much for your review on this!

**Comment by MrSuperTop** [comment]
> @Loran425 Thank you for the review. I will have a look at the comments later, most likely tomorrow.

I have started looking into the commets. I will need some time to implement the changes.

File: tagstudio/src/qt/widgets/preview_panel.py [hunk 4]
```diff
@@ -2,56 +2,51 @@
...
     QMessageBox,
 )
-from humanfriendly import format_size
-
+from src.core.constants import IMAGE_TYPES, RAW_IMAGE_TYPES, VIDEO_TYPES
 from src.core.enums import SettingItems, Theme
 from src.core.library import Entry, ItemType, Library
-from src.core.constants import VIDEO_TYPES, IMAGE_TYPES, RAW_IMAGE_TYPES, TS_FOLDER_NAME
+from src.core.logging import get_logger
+from src.core.constants import TS_FOLDER_NAME
```
[the remark this answers is not in the document]
Comment (MrSuperTop) [reply tagstudio/src/qt/widgets/preview_panel.py:33]: Must have been a mistake of mine when resolving conflicts

**Comment by MrSuperTop** [comment]
> > Definitely want @CyanVoxel's input on the match/case in tag_studio.py and the possibility of removing the `ui` folder, and I think there's some room to discuss the log formatter and what information should be displayed as well as some questions on the ContextVars and splitting out the arg parser, though that split has been growing on me.
> 
> It would probably be good to scrap the UI folder at this point. I think the UI development is pretty solidified in just Python at this point, for better or for worse. And I commented above but just in case it gets missed, I'm all for adding in match/case statements!
> 
> Thanks so much for your review on this!

Am I understanding corretly that I can feel free to remove the UI folder as well as the holdover code relating to it? Or would you be willing to handle it in a separate PR?

File: tagstudio/src/core/logging.py
[the remark this answers is not in the document]
Comment (MrSuperTop) [reply tagstudio/src/core/logging.py:1]: > This file should probably be named something other than logging since that shadows the logging module. Might not be an issue since everything imports it as `src.core.logging`

I assume that this shouldn't be an issue as all the imports are global

**Comment by yedpodtrzitko** [comment]
It's generally a good practice to keep the changes relevant to the topic - if the title says it's about logging refactoring, then most likely it shouldnt contain a refactoring of argparser,  or adding new ruff checks and changes related to it. 
Also if the `ruff check` will be considered compulsory, then it should be added into the CI pipeline as well, otherwise the next commit from someone who doesnt use pre-commit will introduce regression there.

**Comment by yedpodtrzitko** [comment]
on logging-relevant note. A nice improvement I could imagine would be using [structlog](https://github.com/hynek/structlog/), which is a drop-in replacement with nice formatting out of the box minus all the extra work like manually prefixing strings via `{INFO}` or `{ERROR}`, which is quite easy to forget to do.

It also allows to pass the variables as actual variables instead of mashing everything into an f-string. So for example this:
```
logger.info(f"Query:{query}, Frame: {i},  Length: {len(f)}")
```
can be written as this:
```
logger.info("Search Results", query=query, frame=i, length=len(f))
```

output looks like this:

Screenshot 2024-06-09 at 20 59 02

File: tagstudio/src/cli/ts_cli.py
[the remark this answers is not in the document]
Comment (MrSuperTop) [reply tagstudio/src/cli/ts_cli.py:1]: > I don't believe CLI is currently active so I just gave this a cursory look with nothing jumping out at me but there may be something I missed

So do you still maintain the `CliDriver`? If not, why is it still present in the codebase?

File: tagstudio/parse_args.py
[the remark this answers is not in the document]
Comment (MrSuperTop) [reply tagstudio/parse_args.py:1]: > Not entirely sure the benefit of splitting out the argument parser into its own file unless its anticipated that other sections of the code will be referencing command line options. having this be its own function does make the `main` function a touch easier to read but currently the parsed args are passed to the driver rather than the drivers needing to refence the args directly.
> 
> I think the caching makes a touch more sense if the idea was that other files would request the args directly.
> 
> Not saying this needs to change from your code necessarily just trying to understand the reasoning.

I would say, that in this particular version of the code there is no concrete reason to split out the argument parser into a separate file, I simply did it as a habbit and it seemed more natural to have the logic in a separate file.

However, I have also thought about subclassing `argparse.Namespace` and creating a custom namespace to enable typechecking as well as solidify the arguments namespace signature. In this case, I think that putting the namespace definition in the same file as the argument parsing logic is reasonable and it is a solid argument for splitting everything into a separate file.

I have already implemented the custom namespace, I will commit the changes shortly together with other changes in response to the comments.

File: tagstudio/src/core/logging.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import logging
+from contextvars import ContextVar
+
+_root_logger = logging.getLogger()
+_main_logger = _root_logger.getChild("tag_studio")
+
+_main_logger_context_var: ContextVar[logging.Logger] = ContextVar("main_logger")
+_main_logger_context_var.set(_main_logger)
+
```
[the remark this answers is not in the document]
Comment (MrSuperTop) [reply tagstudio/src/core/logging.py:9]: > I'm guessing this is a thread safety feature since normally `getLogger(__name__)` will return the same logger reference without needing a global or contextvar
> 
> Forgive me if this is a common practice, I just haven't run across ContextVars before

The primary reason as to why a simple `getLogger(__name__)` was not used in the `get_logger` function is that in my opinion having some kind of "main logger" is logical in this particular instance, due to how the `__name__` variable won't contain the "tag_studio" part as tag_studio is not a package in the first place.

I am not certain whether the usage of contextvars for this purpose is a common practice anywhere, as I don't have much experience with real codebases. I have used a contextvar here in order to have a common "main logger" inside the execution context. I assume, that I could have simply used a top-level variable instead of a contextvar in this particular instance, and it wouldn't make any difference. There is no concrete reason as to why a contextvar is better here, it simply introduces unnecessary complexity and additional overhead. I could imagine a situation where someone would want different "main logger" for some parts of their system or subsystem but this is definetely not the case here. Would you suggest that I should use the already defined `_main_logger` variable inside the `get_logger` function and to get rid of the contextvar?

Regarding the thread safety, as far as I have understood after briedfly skimming the documentation and doing a small test, each thread will have it's own `contextvars.Context`, so the contextvar will introduce issues if someone where to try to call `get_logger` from a different thread. On the other hand, `logging.Logger.getChild` should be thread-safe and return the same `logging.Logger` object, provided the name is the same.

File: tagstudio/src/qt/widgets/preview_panel.py [hunk 6]
```diff
@@ -835,7 +830,12 @@ def write_container(self, index, field, mixed=False):
...
-                callback = lambda: (self.remove_field(field), self.update_widgets())
+
+                # TODO: Reuse the callback
+                # NOTE: Not a lambda as it was previously due to ruff complaining about lambdas being assigned to variables
+                def callback():
+                    return self.remove_field(field), self.update_widgets()
+
                 container.set_remove_callback(
                     lambda: self.remove_message_box(prompt=prompt, callback=callback)
                 )
```
[the remark this answers is not in the document]
Comment (MrSuperTop) [reply tagstudio/src/qt/widgets/preview_panel.py:841]: So, should I go ahead and change this to use a partial or is this beyond the scope of this PR? This change in the first place was made by `ruff check --fix`.

[the remark this answers is not in the document]
Comment (Loran425) [reply tagstudio/src/qt/widgets/preview_panel.py:841]: I would leave the partials for another PR. 
There are other behavioral changes that need to take place such as having `self.update_widgets` be called by a signal. 

So for now I would leave the callback function if that's making ruff happy and then someone can look at the partials and adding a signal for widget updates.

File: tagstudio/src/cli/ts_cli.py
[the remark this answers is not in the document]
Comment (Loran425) [reply tagstudio/src/cli/ts_cli.py:1]: I haven't touched the CLI, I believe @CyanVoxel had it as an older version that was just part of the original upload and we've all been so busy chasing other things that we haven't considered removing it. I think there has been at least one request for a CLI so it might make sense to keep but the primary focus has been on the Qt(python)/Tauri(rust) front ends.

[the remark this answers is not in the document]
Comment (CyanVoxel) [reply tagstudio/src/cli/ts_cli.py:1]: Before open-sourcing the program, the CLI used to be the only functional and feature-complete frontend, initially created to test out the functionality of the newly-refactored codebase. While there was a brief span of time where I was working to maintain both the CLI and Qt GUI frontends, I quickly shifted focus towards the Qt version to get it at least up to parity. Fast forward to when I open-sourced the codebase in the state is was, and there was still a few things that the (now depreciated) CLI was capable of that the GUI wasn't *(and still isn't - including deleting tags, configuring collages, and resolving multi-matched unlinked entries).*

TLDR I've kept the CLI file hanging around as a resource for how certain features can be implemented, even if the CLI frontend itself has remained unmaintained. Training wheels for the Qt version and the repo as a whole essentially. I've been considering cutting it from the repo soon or at the very least keeping it in another branch in case anyone has the desire to bring it up to spec and maintain it.

**Comment by Loran425** [comment]
> > > Definitely want CyanVoxel's input on the match/case in tag_studio.py and the possibility of removing the `ui` folder, and I think there's some room to discuss the log formatter and what information should be displayed as well as some questions on the ContextVars and splitting out the arg parser, though that split has been growing on me.
> > 
> > 
> > It would probably be good to scrap the UI folder at this point. I think the UI development is pretty solidified in just Python at this point, for better or for worse. And I commented above but just in case it gets missed, I'm all for adding in match/case statements!
> > Thanks so much for your review on this!
> 
> Am I understanding corretly that I can feel free to remove the UI folder as well as the holdover code relating to it? Or would you be willing to handle it in a separate PR?

Since the addition of `ruff check` won't be happy with those lines not changing I would delete the commented lines and a separate PR can clean the files up if you want to make that as the files will be unreferenced and unused files at that point. 

Unless this is getting broken up into 3 or so PR's 1 for arg parse, 1 for ruff, 1 for logging. in which case lots of changes would be needed. 

<hr>

 > It's generally a good practice to keep the changes relevant to the topic - if the title says it's about logging refactoring, then most likely it shouldnt contain a refactoring of argparser, or adding new ruff checks and changes related to it. Also if the `ruff check` will be considered compulsory, then it should be added into the CI pipeline as well, otherwise the next commit from someone who doesnt use pre-commit will introduce regression there.

@yedpodtrzitko that's just a tweak to the ruff work flow? or it needs a whole new workflow? 
Modified `.github\workflows\ruff.yaml`
```yaml 
name: Ruff
on: [ push, pull_request ]
jobs:
  ruff-format:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: chartboost/ruff-action@v1
        with:
          version: 0.4.2
          args: 'format --check'   
      - uses: chartboost/ruff-action@v1  #uses `check` arg by default
        with:
          version: 0.4.2
```

**Comment by yedpodtrzitko** [comment]
> @yedpodtrzitko that's just a tweak to the ruff work flow? or it needs a whole new workflow? Modified `.github\workflows\ruff.yaml`

modifying the existing workflow like this should be sufficient

**Comment by MrSuperTop** [comment]
> Since the addition of ruff check won't be happy with those lines not changing I would delete the commented lines and a separate PR can clean the files up if you want to make that as the files will be unreferenced and unused files at that point.
> Unless this is getting broken up into 3 or so PR's 1 for arg parse, 1 for ruff, 1 for logging. in which case lots of changes would be needed.

I would be willing to make a separate PR to add the ruff check pre-commit hook as well as change the workflow. Should I also include the mypy pre-commit hook? As there is a CI workflow and isn't a pre-commit hook, so the only way to know whether mypy fails is to commit and then run the workflow.

File: tagstudio/src/qt/widgets/preview_panel.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MrSuperTop) [reply tagstudio/src/qt/widgets/preview_panel.py:841]: I have added a TODO comment near one of the callback definitions referencing the issue. I suppose this change is beyond the scope of this PR. I will mark this as resolved.

**Comment by CyanVoxel** [comment]
Closing this as the codebase has changed significantly since this last had any activity.
Thank you all for the valuable discussions here that will help inform future PRs.
